aws-opentelemetry-distro/src/amazon/opentelemetry/distro/_aws_metric_attribute_generator.py

```diff
@@ -344,28 +353,100 @@ def _set_remote_type_and_identifier(span: ReadableSpan, attributes: BoundedAttri
+            )
+    elif is_db_span(span):
+        remote_resource_type = _DB_CONNECTION_STRING_TYPE
+        remote_resource_identifier = _get_db_connection(span)
 
     if remote_resource_type is not None and remote_resource_identifier is not None:
         attributes[AWS_REMOTE_RESOURCE_TYPE] = remote_resource_type
         attributes[AWS_REMOTE_RESOURCE_IDENTIFIER] = remote_resource_identifier
 
 
+def _get_db_connection(span: ReadableSpan) -> None:
+    """
+    RemoteResourceIdentifier is populated with rule:
+        ^[{db.name}|]?{address}[|{port}]?
+
+    {address} attribute is retrieved in priority order:
+    - {SpanAttributes.SERVER_ADDRESS},
+    - {SpanAttributes.NET_PEER_NAME},
+    - {SpanAttributes.SERVER_SOCKET_ADDRESS},
+    - {SpanAttributes.DB_CONNECTION_STRING}-Hostname
+
+    {port} attribute is retrieved in priority order:
+    - {SpanAttributes.SERVER_PORT},
+    - {SpanAttributes.NET_PEER_PORT},
+    - {SpanAttributes.SERVER_SOCKET_PORT},
+    - {SpanAttributes.DB_CONNECTION_STRING}-Port
+
+    If address is not present, neither RemoteResourceType nor RemoteResourceIdentifier will be provided.
+    """
+    db_name: Optional[str] = span.attributes.get(_DB_NAME)
+    db_connection: Optional[str] = None
+
+    if is_key_present(span, _SERVER_ADDRESS):
+        server_address: Optional[str] = span.attributes.get(_SERVER_ADDRESS)
+        server_port: Optional[str] = span.attributes.get(_SERVER_PORT)
+        db_connection = _build_db_connection(server_address, server_port)
+    elif is_key_present(span, _NET_PEER_NAME):
+        network_peer_address: Optional[str] = span.attributes.get(_NET_PEER_NAME)
+        network_peer_port: Optional[str] = span.attributes.get(_NET_PEER_PORT)
+        db_connection = _build_db_connection(network_peer_address, network_peer_port)
+    elif is_key_present(span, _SERVER_SOCKET_ADDRESS):
+        server_socket_address: Optional[str] = span.attributes.get(_SERVER_SOCKET_ADDRESS)
+        server_socket_port: Optional[str] = span.attributes.get(_SERVER_SOCKET_PORT)
+        db_connection = _build_db_connection(server_socket_address, server_socket_port)
+    elif is_key_present(span, _DB_CONNECTION_STRING):
+        connection_string: Optional[str] = span.attributes.get(_DB_CONNECTION_STRING)
+        db_connection = _build_db_connection_string(connection_string)
+
+    if db_connection and db_name:
+        db_connection = _escape_delimiters(db_name) + "|" + db_connection
+
+    return db_connection
+
+
+def _build_db_connection(address: str, port: int) -> Optional[str]:
+    return _escape_delimiters(address) + ("|" + str(port) if port else "")
+
+
+def _build_db_connection_string(connection_string: str) -> Optional[str]:
+    uri = urlparse(connection_string)
+    address = uri.hostname
+    port = uri.port
+
+    if address is None:
+        return None
+
+    port_str = "|" + str(port) if port is not None and port != -1 else ""
```

Comment:
Java and python do not handle invalid ports the same here: https://docs.python.org/3/library/urllib.parse.html

> Changed in version 3.6: Out-of-range port numbers now raise [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError), instead of returning [None](https://docs.python.org/3/library/constants.html#None).

Need to wrap in try/catch.

Reply:
updated.